### PR TITLE
ARROW-6000: [Python] Add support for LargeString and LargeBinary types 

### PR DESCRIPTION
**Pull request by pitrou**
Also fix a bug in Take / Filter for large binary types.

### CODE REVIEW

**Comment by kszucs** [comment]
@ursabot build

**Comment by pitrou** [comment]
Travis-CI: https://travis-ci.org/pitrou/arrow/builds/565380024

**Comment by pitrou** [comment]
@jorisvandenbossche Want to take a look?

**Comment by jorisvandenbossche** [comment]
Yes, was planning to do that, thanks for the ping!

I played a bit with the branch (mainly with large_string), some things I noticed (to be clear: I suppose most of them are expected to not yet working, so not wanting to say they should be fixed in this PR, but some items might be worth JIRA issues):

* With `dictionary_encode()` is not working on the array (`a = pa.array(['a', 'b', 'c'], pa.large_string()); a.dictionary_encode()` gives NotImplementedError), but constructing a dictionary array manually with large string dictionary works OK. `unique` is also not implemented.

* `to_pandas` is not yet working (both on Table or Array). `to_pylist` is working fine. Also not yet for parquet, but I don't know for sure that parquet has a type that could be used for this (so not sure it should be working).

* Do we want to be able to cast string to large_string (or the other way around) ?

* `take` is not working correctly (but also not raising that it is not supported), while it works correctly for normal string:

    ```
    In [29]: a = pa.array(['a', 'b', 'c'], pa.large_string()) 

    In [30]: a.take(pa.array([0, 2])) 
    Out[30]: 
    <pyarrow.lib.LargeStringArray object at 0x7f19719d40f8>
    [
      "",
      ""
    ]

    In [31]: b = pa.array(['a', 'b', 'c'], pa.string())

    In [32]: b.take(pa.array([0, 2]))
    Out[32]: 
    <pyarrow.lib.StringArray object at 0x7f19719cd2b0>
    [
      "a",
      "c"
    ]
    ```

Will take a look at the actual code now.

**Comment by pitrou** [comment]
Ok, some answers:
* dictionary encoding isn't implemented, that's expected.
* `to_pandas`: similar. Perhaps that's something you would like to tackle in a separate issue?

**Comment by jorisvandenbossche** [comment]
Ah, I see you already opened https://issues.apache.org/jira/browse/ARROW-6071 for the casting yesterday

> to_pandas: similar. Perhaps that's something you would like to tackle in a separate issue?

Yes, that's something I could look into.

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -1531,6 +1532,33 @@ def binary(int length=-1):
     return pyarrow_wrap_data_type(fixed_size_binary_type)
 
 
+def large_binary():
+    """
+    Create large variable-length binary type
+
+    This data type may not be supported by all Arrow implementations.  Unless
+    you need to represent data larger than 2GB, you should prefer binary().
```
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi]: should we clarify this is about the total size of a single array?

**Review by jorisvandenbossche — COMMENTED** [review]
I already went through the python part (need to run now), which looks good.

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/types.pxi]: Ah, forget, a single value bigger than 2GB of course implies an array that is bigger than 2GB .. ;), so the distinction does not matter much.

**Comment by pitrou** [comment]
Ok, I've pushed a fix for the `take()` issue. Thanks for noticing!

**Comment by pitrou** [comment]
Will merge soon if no more comments.

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -612,6 +612,7 @@ class ARROW_EXPORT StringType : public BinaryType {
  public:
   static constexpr Type::type type_id = Type::STRING;
   static constexpr bool is_utf8 = true;
+  using EquivalentBinaryType = BinaryType;
```
Comment (pitrou) [inline cpp/src/arrow/type.h]: @fsaintjacques Any issue with the naming?

**Comment by jorisvandenbossche** [comment]
> to_pandas: similar. Perhaps that's something you would like to tackle in a separate issue?

Have to see a bit with my priorities, but at least created an issue for it: https://issues.apache.org/jira/browse/ARROW-6115

**Pull request closed without merging**